File: modem.js
const merge = require('lodash.merge');
const http = require('http');
const qs = require('qs');
const { hex_md5 } = require('./md5');
const {
  getCurrentTimeString,
  encodeMessage,
  decodeMessage,
  transTime,
  promiseRetry,
} = require('./utils');

class Modem {
  constructor({ modemIP = '192.168.0.1', modemPassword = ''} = {}) {
    this.modemIP = modemIP;
    this.modemPassword = modemPassword
    this.getPath = `/goform/goform_get_cmd_process`;
    this.setPath = `/goform/goform_set_cmd_process`;
    this.referer = `http://${this.modemIP}/index.html`;

    this.loginCookieValue = '';
    this.modemVersion = '';
  }

  async #request(options, data = '') {
    const headers = { Referer: this.referer };

    data = qs.stringify(data);
    options.path = this.getPath;
    if (options.method === 'POST') {
      options.path = this.setPath;
      headers['Content-Length'] = data.length;
    } else if (options.method === 'GET') {
      options.path += `?${data}`;
      data = '';
    }

    options = merge(options, {
      hostname: this.modemIP,
      insecureHTTPParser: true,
      headers,
    });

    return new Promise((resolve, reject) => {
      const request = http.request(options, (response) => {
        let data = '';
        response.on('data', (chunk) => {
          data += chunk.toString();
        });
        response.on('end', () => {
          resolve({ response, data: JSON.parse(data) });
        });
        response.on('error', (error) => {
          reject(error);
        });
      });
      request.on('error', (error) => { reject(error); });
      data && request.write(data);
      request.end();
    });
  }

  async #login() {
    const data = {
      isTest: false,
      goformId: 'LOGIN',
      password: Buffer.from(this.modemPassword).toString('base64'),
    };
    const options = { method: 'POST' };
    const response = await this.#request(options, data);
    if (response.data.result !== '0') {
      throw new Error('Login to modem failed.');
    }
    this.loginCookieValue = response.response.headers['set-cookie'][0].split(';')[0];
  }

  async #logout() {
    const data = {
      isTest: false,
      goformId: 'LOGOUT',
      AD: await this.#getAD(),
    };
    const options = { method: 'POST' };
    await this.#request(options, data);
    this.loginCookieValue = '';
  }

  async #getModemVersion() {
    if (this.modemVersion) {
      return this.modemVersion;
    }
    const data = {
      isTest: false,
      cmd: 'cr_version,wa_inner_version',
      multi_data: 1,
    }
    const options = { method: 'GET' };
    const response = await this.#request(options, data);
    if (response.data.wa_inner_version || response.data.cr_version) {
      this.modemVersion = `${response.data?.cr_version}${response.data?.wa_inner_version}`;
      return this.modemVersion;
    } else {
      throw new Error('Getting modem version failed.');
    }
  }

  async #getAD() {
    const modemVersion = await this.#getModemVersion();
    const RD = await this.#getRD();
    return hex_md5(`${hex_md5(modemVersion)}${RD}`);
  }

  async #getRD() {
    const data = {
      isTest: false,
      cmd: 'RD',
    }
    const options = { method: 'GET' };
    const response = await this.#request(options, data);

    if (!response.data.RD) {
      throw new Error('Getting RD failed.');
    }
    return response.data.RD;
  }

  async #getCmdStatusInfo(sms_cmd) {
    const data = {
      isTest: false,
      cmd: 'sms_cmd_status_info',
      sms_cmd,
    };
    const options = {
      method: 'GET',
      headers: { Cookie: this.loginCookieValue },
    };
    const response = await this.#request(options, data);
    return response.data;
  }

  async #awaitConfirmation(sms_cmd) {
    await promiseRetry(async () => {
      const info = await this.#getCmdStatusInfo(sms_cmd);
      return (!info?.sms_cmd_status_result || info.sms_cmd_status_result === '1')
        ? Promise.reject()
        : Promise.resolve();
    });
  }

  async getSmsCapacityInfo() {
    await this.#login();
    const data = {
      isTest: false,
      cmd: 'sms_capacity_info',
    };
    const options = {
      method: 'GET',
      headers: { Cookie: this.loginCookieValue },
    };
    const response = await this.#request(options, data);
    await this.#logout();
    return response.data;
  }

  async getAllSms() {
    await this.#login();
    const data = {
      isTest: false,
      cmd: 'sms_data_total',
      page: 0,
      data_per_page: 5000,
      mem_store: 1,
      tags: 10,
      order_by: 'order by id desc',
    };
    const options = {
      method: 'GET',
      headers: { Cookie: this.loginCookieValue },
    };
    const response = await this.#request(options, data);
    await this.#logout();
    response.data.messages.forEach((m) => {
      m.date = transTime(m.date, '3', '24');
      m.content = decodeMessage(m.content);
    });
    return response.data.messages;
  }

  async deleteSms(ids) {
    ids = Array.isArray(ids) ? ids : Array(ids);
    await this.#login();
    const data = {
      isTest: false,
      goformId: 'DELETE_SMS',
      msg_id: `${ids.join(';')};`,
      notCallback: true,
      AD: await this.#getAD(),
    };
    const options = {
      method: 'POST',
      headers: { Cookie: this.loginCookieValue },
    };
    const response = await this.#request(options, data);
    await this.#awaitConfirmation(6);
    await this.#logout();
    if (response.data.result !== 'success') {
      throw new Error('Error deleting SMS.');
    }
  }

  // sms tags
  // 1	Unread received message.
  // 0	Received message.
  // 2	Message sent.
  // 3	Message failed to be sent.
  // 4	Draft.
  async deleteAllSms(tag = '') {
    tag = tag.toString();
    let messages = await this.getAllSms();
    messages = messages.filter((m) => !tag || m.tag === tag);
    const messageIds = messages.map((m) => m.id);
    return await this.deleteSms(messageIds);
  }

  async setSmsAsRead(ids) {
    ids = Array.isArray(ids) ? ids : Array(ids);
    await this.#login();
    const data = {
      isTest: false,
      goformId: 'SET_MSG_READ',
      msg_id: `${ids.join(';')};`,
      tag: '0',
      AD: await this.#getAD(),
    };
    const options = {
      method: 'POST',
      headers: { Cookie: this.loginCookieValue },
    };
    const response = await this.#request(options, data);
    if (response.data.result !== 'success') {
      throw new Error('Error marking SMS as read.');
    }
    await this.#awaitConfirmation(5);
    await this.#logout();
  }

  async setAllSmsAsRead() {
    let messages = await this.getAllSms();
    messages = messages.filter((m) => m.tag === '1');
    const messageIds = messages.map((m) => m.id);
    return await this.setSmsAsRead(messageIds);
  }

  async sendSms(number, message) {
    await this.#login();
    const sms_time = getCurrentTimeString();
    const sms_time_compare = transTime(sms_time.replaceAll(';', ','), '3', '24');
    const data = {
      isTest: false,
      goformId: 'SEND_SMS',
      notCallback: true,
      Number: number.toString(),
      sms_time,
      MessageBody: encodeMessage(message),
      ID: -1,
      encode_type: 'UNICODE',
      AD: await this.#getAD(),
    };
    const options = {
      method: 'POST',
      headers: { Cookie: this.loginCookieValue },
    };
    const response = await this.#request(options, data);
    if (response.data.result !== 'success') {
      throw new Error('Error sending SMS.');
    }
    await this.#awaitConfirmation(4);
    let messages = await this.getAllSms();
    message = messages.filter((m) => 
      m.tag === '2' &&
      m.number === number &&
      m.content === message &&
      sms_time_compare === m.date
    );
    await this.#logout();
    return message;
  }
}

module.exports = Modem;

File: ZteSms/Modem.cs
namespace ZteSms;

using System;
using System.Net.Http;
using System.Threading.Tasks;
using System.Text;
using System.Collections.Generic;
using System.Text.Json;

public class Modem
{
    private readonly HttpClient _httpClient = new();
    private readonly string _modemIP;
    private readonly string _modemPassword;
    private string _loginCookie = string.Empty;
    private string _modemVersion = string.Empty;

    public Modem(string modemIP = "192.168.0.1", string modemPassword = "")
    {
        _modemIP = modemIP;
        _modemPassword = modemPassword;
        _httpClient.BaseAddress = new Uri($"http://{_modemIP}");
        _httpClient.DefaultRequestHeaders.Referrer = new Uri($"http://{_modemIP}/index.html");
    }

    private async Task<HttpResponseMessage> RequestAsync(HttpMethod method, string path, Dictionary<string, string> data)
    {
        FormUrlEncodedContent content = new(data ?? new());
        HttpRequestMessage request = new(method, path);
        if (method == HttpMethod.Get && data is not null)
        {
            request.RequestUri = new Uri($"http://{_modemIP}{path}?{await content.ReadAsStringAsync()}");
        }
        else if (method == HttpMethod.Post)
        {
            request.Content = content;
        }
        if (!string.IsNullOrEmpty(_loginCookie))
        {
            request.Headers.Add("Cookie", _loginCookie);
        }
        return await _httpClient.SendAsync(request);
    }

    private async Task LoginAsync(bool isTest = false)
    {
        Dictionary<string, string> data = new()
        {
            ["isTest"] = isTest.ToString(),
            ["goformId"] = "LOGIN",
            ["password"] = Convert.ToBase64String(Encoding.UTF8.GetBytes(_modemPassword))
        };
        HttpResponseMessage response = await RequestAsync(HttpMethod.Post, "/goform/goform_set_cmd_process", data);
        ResultResponse? json = JsonSerializer.Deserialize<ResultResponse>(await response.Content.ReadAsStringAsync());
        if (json?.result != "0")
        {
            throw new Exception("Login to modem failed.");
        }

        if (response.Headers.TryGetValues("Set-Cookie", out IEnumerable<string>? cookies))
        {
            _loginCookie = cookies.First().Split(';')[0];
        }
    }

    private async Task LogoutAsync(bool isTest = false)
    {
        Dictionary<string, string> data = new()
        {
            ["isTest"] = isTest.ToString(),
            ["goformId"] = "LOGOUT",
            ["AD"] = await GetADAsync()
        };
        await RequestAsync(HttpMethod.Post, "/goform/goform_set_cmd_process", data);
        _loginCookie = string.Empty;
    }

    private async Task<string> GetModemVersionAsync(bool isTest = false)
    {
        if (!string.IsNullOrEmpty(_modemVersion))
        {
            return _modemVersion;
        }

        Dictionary<string, string> data = new()
        {
            ["isTest"] = isTest.ToString(),
            ["cmd"] = "cr_version,wa_inner_version",
            ["multi_data"] = "1"
        };
        HttpResponseMessage response = await RequestAsync(HttpMethod.Get, "/goform/goform_get_cmd_process", data);
        VersionResponse? json = JsonSerializer.Deserialize<VersionResponse>(await response.Content.ReadAsStringAsync());
        string crStr = json?.cr_version ?? string.Empty;
        string waStr = json?.wa_inner_version ?? string.Empty;
        if (!string.IsNullOrEmpty(crStr) || !string.IsNullOrEmpty(waStr))
        {
            _modemVersion = $"{crStr}{waStr}";
            return _modemVersion;
        }
        throw new Exception("Getting modem version failed.");
    }

    private static string HexMD5(string input)
    {
        byte[] bytes = System.Security.Cryptography.MD5.HashData(Encoding.UTF8.GetBytes(input));
        StringBuilder sb = new();
        foreach (byte b in bytes)
        {
            sb.Append(b.ToString("x2"));
        }

        return sb.ToString();
    }

    private async Task<string> GetRDAsync(bool isTest = false)
    {
        Dictionary<string, string> data = new()
        {
            ["isTest"] = isTest.ToString(),
            ["cmd"] = "RD"
        };
        HttpResponseMessage response = await RequestAsync(HttpMethod.Get, "/goform/goform_get_cmd_process", data);
        RdResponse? json = JsonSerializer.Deserialize<RdResponse>(await response.Content.ReadAsStringAsync());
        if (string.IsNullOrEmpty(json?.RD))
        {
            throw new Exception("Getting RD failed.");
        }

        return json.RD;
    }

    private async Task<string> GetADAsync()
    {
        string version = await GetModemVersionAsync();
        string rd = await GetRDAsync();
        return HexMD5(HexMD5(version) + rd);
    }

    private async Task AwaitConfirmationAsync(int cmd)
    {
        for (int i = 0; i < 20; i++)
        {
            CmdStatusInfo? info = await GetCmdStatusInfoAsync(cmd);
            if (!string.IsNullOrEmpty(info?.sms_cmd_status_result) && info.sms_cmd_status_result != "0")
            {
                return;
            }

            await Task.Delay(200);
        }
        throw new Exception("Command confirmation timeout.");
    }

    private async Task<CmdStatusInfo?> GetCmdStatusInfoAsync(int cmd, bool isTest = false)
    {
        Dictionary<string, string> data = new()
        {
            ["isTest"] = isTest.ToString(),
            ["cmd"] = "sms_cmd_status_info",
            ["sms_cmd"] = cmd.ToString()
        };
        HttpResponseMessage response = await RequestAsync(HttpMethod.Get, "/goform/goform_get_cmd_process", data);
        return JsonSerializer.Deserialize<CmdStatusInfo>(await response.Content.ReadAsStringAsync());
    }

    public async Task<SmsCapacityInfo?> GetSmsCapacityInfoAsync(bool isTest = false)
    {
        await LoginAsync();
        Dictionary<string, string> data = new()
        {
            ["isTest"] = isTest.ToString(),
            ["cmd"] = "sms_capacity_info"
        };
        HttpResponseMessage response = await RequestAsync(HttpMethod.Get, "/goform/goform_get_cmd_process", data);
        await LogoutAsync();
        return JsonSerializer.Deserialize<SmsCapacityInfo>(await response.Content.ReadAsStringAsync());
    }

    public async Task<List<Message>> GetAllSmsAsync(bool isTest = false)
    {
        await LoginAsync();
        Dictionary<string, string> data = new()
        {
            ["isTest"] = isTest.ToString(),
            ["cmd"] = "sms_data_total",
            ["page"] = "0",
            ["data_per_page"] = "5000",
            ["mem_store"] = "1",
            ["tags"] = "10",
            ["order_by"] = "order by id desc"
        };
        HttpResponseMessage response = await RequestAsync(HttpMethod.Get, "/goform/goform_get_cmd_process", data);
        await LogoutAsync();
        AllSmsResponse? json = JsonSerializer.Deserialize<AllSmsResponse>(await response.Content.ReadAsStringAsync());
        return json?.messages ?? new List<Message>();
    }

    public async Task DeleteSmsAsync(IEnumerable<string> ids, bool isTest = false)
    {
        await LoginAsync();
        Dictionary<string, string> data = new()
        {
            ["isTest"] = isTest.ToString(),
            ["goformId"] = "DELETE_SMS",
            ["msg_id"] = string.Join(';', ids) + ";",
            ["notCallback"] = "true",
            ["AD"] = await GetADAsync()
        };
        HttpResponseMessage response = await RequestAsync(HttpMethod.Post, "/goform/goform_set_cmd_process", data);
        await AwaitConfirmationAsync(6);
        await LogoutAsync();
        ResultResponse? json = JsonSerializer.Deserialize<ResultResponse>(await response.Content.ReadAsStringAsync());
        if (json?.result != "success")
        {
            throw new Exception("Error deleting SMS.");
        }
    }

    public async Task DeleteAllSmsAsync(MessageTag? tag = null)
    {
        List<Message> messages = await GetAllSmsAsync();
        List<string> ids = new();
        foreach (Message msg in messages)
        {
            if (tag is null || msg.Tag == tag)
            {
                ids.Add(msg.id);
            }
        }
        await DeleteSmsAsync(ids);
    }

    public async Task SetSmsAsReadAsync(IEnumerable<string> ids, bool isTest = false)
    {
        await LoginAsync();
        Dictionary<string, string> data = new()
        {
            ["isTest"] = isTest.ToString(),
            ["goformId"] = "SET_MSG_READ",
            ["msg_id"] = string.Join(';', ids) + ";",
            ["tag"] = MessageTag.Received.ToString(),
            ["AD"] = await GetADAsync()
        };
        HttpResponseMessage response = await RequestAsync(HttpMethod.Post, "/goform/goform_set_cmd_process", data);
        ResultResponse? json = JsonSerializer.Deserialize<ResultResponse>(await response.Content.ReadAsStringAsync());
        if (json?.result != "success")
        {
            throw new Exception("Error marking SMS as read.");
        }

        await AwaitConfirmationAsync(5);
        await LogoutAsync();
    }

    public async Task SetAllSmsAsReadAsync()
    {
        List<Message> messages = await GetAllSmsAsync();
        List<string> ids = new();
        foreach (Message msg in messages)
        {
            if (msg.Tag == MessageTag.UnreadReceived)
            {
                ids.Add(msg.id);
            }
        }
        await SetSmsAsReadAsync(ids);
    }

    public async Task<Message> SendSmsAsync(string number, string message, bool isTest = false)
    {
        await LoginAsync();
        string smsTime = DateTime.UtcNow.ToString("yy;MM;dd;HH;mm;ss;+0");
        Dictionary<string, string> data = new()
        {
            ["isTest"] = isTest.ToString(),
            ["goformId"] = "SEND_SMS",
            ["notCallback"] = "true",
            ["Number"] = number,
            ["sms_time"] = smsTime,
            ["MessageBody"] = EncodeMessage(message),
            ["ID"] = "-1",
            ["encode_type"] = "UNICODE",
            ["AD"] = await GetADAsync()
        };
        HttpResponseMessage response = await RequestAsync(HttpMethod.Post, "/goform/goform_set_cmd_process", data);
        ResultResponse? json = JsonSerializer.Deserialize<ResultResponse>(await response.Content.ReadAsStringAsync());
        if (json?.result != "success")
        {
            throw new Exception("Error sending SMS.");
        }

        await AwaitConfirmationAsync(4);
        List<Message> messages = await GetAllSmsAsync();
        foreach (Message msg in messages)
        {
            if (msg.Tag == MessageTag.Sent &&
                msg.number == number &&
                msg.content == message)
            {
                await LogoutAsync();
                return msg;
            }
        }
        await LogoutAsync();
        throw new Exception("Sent SMS not found.");
    }

    private static string EncodeMessage(string message)
    {
        if (string.IsNullOrEmpty(message))
        {
            return string.Empty;
        }

        StringBuilder sb = new();
        foreach (Rune codepoint in message.EnumerateRunes())
        {
            string hex = codepoint.Value.ToString("X");
            sb.Append(hex.PadLeft(4, '0'));
        }
        return sb.ToString();
    }
}
